### PR TITLE
Add layout rule for [[no_unique_address]] (C++20 P0840R2)

### LINKED ISSUE #49: class layout for `[[no_unique_address]]`
C++20 adds a `[[no_unique_address]]` attribute, which allows EBO layout to be requested for non-static data members. We need to update the ABI document to describe how it affects class layout.

As a concrete goal, we should aim to ensure that these two classes are laid out the same:

```
struct A : T1, T2, ... { ... };
struct A { [[no_unique_address]] T1 t1; [[no_unique_address]] T2 t2; ... };
```

... in all cases where that is possible. (There are cases where it is not: for example, if `A` has a primary base class other than `T1` and any prior base class is non-empty.)

### PR DESCRIPTION
**Pull request by zygoloid**
Fixes #49

### CODE REVIEW

**Comment by zygoloid** [comment]
For exposition: the idea here is to use dsize(C) as the allocated size for potentially-overlapping data members in the same way we use nvsize(C) as the allocated size for base classes. We previously only used dsize(C) as an intermediate value when laying out C, and now becomes part of the data propagated to other classes' layouts, but its computed value turns out to already exactly what we want here [except in the case where we aren't allowed to / choose not to reuse tail padding (the finalization step), which is adjusted to also bump dsize(C) to sizeof(C) where needed].

**Comment by rjmccall** [comment]
@zygoloid, would you mind checking my merge and verifying that this change still reflects the right rule according to the standard?

File: abi.html [hunk 1]
```diff
@@ -164,10 +164,15 @@ <h3><a href="#definitions"> 1.1 Definitions </a></h3>
 <dt> <i>empty class</i> </dt>
 <dd>
 A class with no non-static data members,
-no unnamed bit-fields other than zero-width bit-fields,
+no unnamed bit-fields other than zero-width bit-fields and empty data members,
```
Comment (jicama) [inline abi.html]: Shouldn't the "empty data members" be a qualification on "no non-static data members" rather than "no unnamed bit-fields"?

Comment (rjmccall) [reply abi.html]: Yes, absolutely; that was a mistake introduced by my merge.

**Comment by zygoloid** [comment]
Merged rule now looks good, thanks.

**Comment by rjmccall** [comment]
Okay.  I'll leave this open for a week for further comment.

**Comment by rjmccall** [comment]
1 week, 4 weeks, eh.

**Pull request merged**